src/Chainweb/Cut.hs

```diff
@@ -427,6 +428,41 @@ limitCut wdb h c
         -- this is safe because it's guaranteed that the requested rank is
         -- smaller then the block height of the argument
 
+-- | Find a `Cut` that is a predecessor of the given one, and that has a block
+-- height that is as low as possible while not exceeding the given height and
+-- including all of the chains in the given cut.
+--
+-- If the requested limit is larger or equal to the current height, the given
+-- cut is returned.
+--
+tryLimitCut
+    :: HasCallStack
+    => WebBlockHeaderDb
+    -> BlockHeight
+        -- upper bound for the block height of each chain. This is not a tight bound.
+    -> Cut
+    -> IO Cut
+tryLimitCut wdb h c
+    | all (\bh -> h >= _blockHeight bh) (view cutHeaders c) =
+        return c
+    | otherwise = do
+        hdrs <- itraverse go $ view cutHeaders c
+        return $! set unsafeCutHeaders hdrs c
+  where
+    v = _chainwebVersion wdb
+    go :: ChainId -> BlockHeader -> IO BlockHeader
+    go cid bh = do
+        if h >= _blockHeight bh
+        then return bh
+        else do
+            !db <- getWebBlockHeaderDb wdb cid
+            -- this is safe because it's guaranteed that the requested rank is
+            -- smaller then the block height of the argument
+            let ancestorHeight = min (int $ _blockHeight bh) (int h)
+            if ancestorHeight <= fromIntegral (genesisHeight v cid)
```

Comment:
nit: genesisHeight uses genesisBlockHeader, so you can do some CSE and do `let genesis = genesisBlockHeader v cid` and then change this to `if ancestorHeight <= fromIntegral (_blockHeight genesis)`, and the call to genesisBlockHeader below can just be replaced with `genesis`.